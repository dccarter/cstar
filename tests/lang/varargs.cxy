func varargs(...args: auto) {
}

func add(...nums: f32) {
    var sum : f32 = 0;
    #const i = 0;
    #for (const num: nums) {
        sum += nums.#{i}
        #{i += 1}
    }
}

func many(x: i32, ...rest: f32) {
    add(<f32>x, ...rest)
}

func main(args: [string])
{
    varargs(10, "Hello", 'c', 2000.6)
    many(10, 10.0, 100.6)
}