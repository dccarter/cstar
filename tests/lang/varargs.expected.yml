Program:
decls:
- GenericDecl:
  flags: (Variadic|TopLevelDecl)
  decl:
    FuncDecl:
    flags: (Variadic|TopLevelDecl)
    name: varargs
    params:
    - FuncParam:
      flags: (Variadic)
      name: args
      index: 0
      paramType:
        Path:
        flags: (TypeAst)
        elements:
        - PathElem:
          flags: (TypeAst)
          name: _Variadic
        isType: false
    body:
      BlockStmt:
  params:
  - GenericParam:
    flags: (Variadic)
    name: _Variadic
    constraints:
    - AutoType:
      flags: (TypeAst)
- GenericDecl:
  flags: (Variadic|TopLevelDecl)
  decl:
    FuncDecl:
    flags: (Variadic|TopLevelDecl)
    name: add
    params:
    - FuncParam:
      flags: (Variadic)
      name: nums
      index: 0
      paramType:
        Path:
        flags: (TypeAst)
        elements:
        - PathElem:
          flags: (TypeAst)
          name: _Variadic
        isType: false
    body:
      BlockStmt:
  params:
  - GenericParam:
    flags: (Variadic)
    name: _Variadic
    constraints:
    - PrimitiveType:
      flags: (TypeAst)
      name: f32
- GenericDecl:
  flags: (Variadic|TopLevelDecl)
  decl:
    FuncDecl:
    flags: (Variadic|TopLevelDecl)
    name: many
    params:
    - FuncParam:
      name: x
      index: 0
      paramType:
        PrimitiveType:
        flags: (TypeAst)
        name: i32
    - FuncParam:
      flags: (Variadic)
      name: rest
      index: 0
      paramType:
        Path:
        flags: (TypeAst)
        elements:
        - PathElem:
          flags: (TypeAst)
          name: _Variadic
        isType: false
    body:
      BlockStmt:
      stmts:
      - ExprStmt:
        expr:
          CallExpr:
          callee:
            Path:
            elements:
            - PathElem:
              name: add
            isType: false
          args:
          - CastExpr:
            to:
              PrimitiveType:
              flags: (TypeAst)
              name: f32
            exp:
              Path:
              elements:
              - PathElem:
                name: x
              isType: false
          - SpreadExpr:
            expr:
              Path:
              elements:
              - PathElem:
                name: rest
              isType: false
  params:
  - GenericParam:
    flags: (Variadic)
    name: _Variadic
    constraints:
    - PrimitiveType:
      flags: (TypeAst)
      name: f32
- FuncDecl:
  flags: (Variadic|TopLevelDecl)
  name: varargs_0
  params:
  - FuncParam:
    flags: (Variadic)
    name: args
    index: 0
    paramType:
      Path:
      flags: (TypeAst|Variadic)
      elements:
      - PathElem:
        flags: (TypeAst)
        name: _Variadic
      isType: false
  body:
    BlockStmt:
- FuncDecl:
  flags: (Variadic|TopLevelDecl)
  name: many_1
  params:
  - FuncParam:
    name: x
    index: 0
    paramType:
      PrimitiveType:
      flags: (TypeAst)
      name: i32
  - FuncParam:
    flags: (Variadic)
    name: rest
    index: 0
    paramType:
      Path:
      flags: (TypeAst|Variadic)
      elements:
      - PathElem:
        flags: (TypeAst)
        name: _Variadic
      isType: false
  body:
    BlockStmt:
    stmts:
    - Path:
      flags: (Variadic)
      elements:
      - PathElem:
        name: rest
      isType: false
    - ExprStmt:
      expr:
        CallExpr:
        callee:
          Path:
          flags: (Variadic|TopLevelDecl)
          elements:
          - PathElem:
            name: add_2
          isType: false
        args:
        - TupleExpr:
          flags: (Variadic)
          members:
          - CastExpr:
            to:
              PrimitiveType:
              flags: (TypeAst)
              name: f32
            exp:
              Path:
              elements:
              - PathElem:
                name: x
              isType: false
          - MemberExpr:
            target:
              Path:
              flags: (Variadic)
              elements:
              - PathElem:
                name: rest
              isType: false
            member:
              IntegerLit:
              value: 0
          - MemberExpr:
            target:
              Path:
              flags: (Variadic)
              elements:
              - PathElem:
                name: rest
              isType: false
            member:
              IntegerLit:
              value: 1
- FuncDecl:
  flags: (Variadic|TopLevelDecl)
  name: add_2
  params:
  - FuncParam:
    flags: (Variadic)
    name: nums
    index: 0
    paramType:
      Path:
      flags: (TypeAst|Variadic)
      elements:
      - PathElem:
        flags: (TypeAst)
        name: _Variadic
      isType: false
  body:
    BlockStmt:
- FuncDecl:
  flags: (TopLevelDecl)
  name: main
  params:
  - FuncParam:
    name: args
    index: 0
    paramType:
      ArrayType:
      flags: (TypeAst)
      element:
        StringType:
        flags: (TypeAst)
  body:
    BlockStmt:
    stmts:
    - ExprStmt:
      expr:
        CallExpr:
        callee:
          Path:
          flags: (Variadic|TopLevelDecl)
          elements:
          - PathElem:
            name: varargs_0
          isType: false
        args:
        - TupleExpr:
          flags: (Variadic)
          members:
          - IntegerLit:
            value: 10
          - StringLit:
            value: "Hello"
          - CharLit:
            value: 'c'
          - FloatLit:
            value: 2000.600000
    - ExprStmt:
      expr:
        CallExpr:
        callee:
          Path:
          flags: (Variadic|TopLevelDecl)
          elements:
          - PathElem:
            name: many_1
          isType: false
        args:
        - IntegerLit:
          value: 10
        - TupleExpr:
          flags: (Variadic)
          members:
          - FloatLit:
            value: 10.000000
          - FloatLit:
            value: 100.600000
